File: 2638-count-the-number-of-k-free-subsets/2638-count-the-number-of-k-free-subsets.java
class Solution {
    public long countTheNumOfKFreeSubsets(int[] nums, int k) {
        Arrays.sort(nums);
        List<Integer>[] lists = new ArrayList[k];
        long ans = 1;
        for (int i = 0; i < k; i++) {
            lists[i] = new ArrayList<>();
        }
        for (int n: nums) {
            lists[n % k].add(n);
        }
        for (int i = 0; i < k; i++) {
            long a = 1, b = 2;
            List<Integer> list = lists[i];
            int m = list.size();
            if (m == 0) continue;
            for(int j = m - 2; j > -1; j--) {
                if (list.get(j) + k == list.get(j + 1)) {
                    long c = a + b;
                    a = b;
                    b = c;
                }
                else {
                    a = b;
                    b *= 2;
                }
            }
            ans *= b;
        }
        return ans;
    }
    
}

File: 2638-count-the-number-of-k-free-subsets/2638-count-the-number-of-k-free-subsets.py
class Solution:
    def countTheNumOfKFreeSubsets(self, nums: List[int], k: int) -> int:
        nums.sort()
        lists = [list() for i in range(k)]
        for num in nums:
            lists[num%k].append(num)
        ans = 1
        
        for l in lists:
            m = len(l)
            @cache
            def res(pos = 0):
                if pos == m:
                    return 1
                elif pos == m - 1:
                    return 2
                elif l[pos] + k == l[pos + 1]:
                    return res(pos + 1) + res(pos + 2)
                else:
                    return res(pos + 1) * 2
            ans *= res()
        
        return ans
            